### PR TITLE
To Do list: add & remove

### PR DESCRIPTION
**Pull request by Bestbynature**
### Features included in this pull request are as follows:

- [ ] ⚡ All hardcoded items from the tasks array have been removed. :white_check_mark: 
- [ ] ⚡ A new JavaScript file was created for the new functionalities. :white_check_mark:
- [ ] ⚡ A function for adding a new task is implemented which also adds the particular task to the tasks array. :white_check_mark:
- [ ] ⚡ A function for deleting a task is also implemented and it removes the particular element from the array. :white_check_mark:
- [ ] ⚡ A function for editing task descriptions is implemented. :white_check_mark:
- [ ] ⚡ New tasks have the property `completed` set to `false `and the property `index` set to the value of the new array length. :white_check_mark:
- [ ] ⚡ Deleting a task updates all remaining items' indexes, so they represent the current list order and are unique. :white_check_mark:
- [ ] ⚡ All changes to the To Do List are effectively saved in local storage. :white_check_mark:

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -1,36 +1,144 @@
...
+  localStorage.setItem('tasks', JSON.stringify(tasks));
+};
+
+const displayTasks = () => {
+  tasks.forEach((task, i) => {
+    const taskPane = document.createElement('div');
+    taskPane.className = 'task-pane';
+
+    const topLeft = document.createElement('div');
+    topLeft.className = 'top-left';
```
Comment (thecodechaser) [inline src/index.js]: - [x] You did a great job on the project but you missed one requirement. Please make sure to place the new functionality code in a new JavaScript file as requested in the project requirements. I would suggest creating a directory called `modules` and place the new file in that folder.

image

**Review by thecodechaser — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️ 

## Hi @Bestbynature👋, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## To Highlight 🎉
- Nice UI design. ✔️
- No linter errors. :heavy_check_mark:
- Professional README file. :heavy_check_mark:
- PR has a good title and summary. ✔️
- GitHub flow is followed. :heavy_check_mark:



## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: src/functionality.js [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+      tasks.push(task);
+      todo.value = '';
+      domTasks.innerHTML = '';
+      displayTasks();
+    }
+  } else {
+    abnormal.style.display = 'block';
+    abnormal.innerHTML = 'Sorry! You can\'t add an empty task';
+  }
+};
```
Comment (danmainah) [inline src/functionality.js]: - [ ] @Bestbynature great work trying to work on the changes requested, however, I see you only changed the file name, kindly consider renaming the file index.js and then move the add and delete functionality to a new file which will be imported to the index.js file as previously requested.

**Review by danmainah — CHANGES_REQUESTED** [review]
Hi @Bestbynature , 

Great work on making the changes requested by a previous reviewer 👏  👏 👏
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
✔️ No linter errors
✔️ Github flow was correctly used
✔️  Good design


 # Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by medaminedev66** [comment]
Hello @Bestbynature 👋,

_This is an additional review as a result of the second opinion request._

All changes requested by previous reviewers are part of project requirements. Please check out the requirements below 👇 :

image

image

The first reviewer requested creating a directory called modules and place a new file in that folder. The second reviewer did not find a directory called models, you have only changed the name of the `index.js` file to `functionality`.

As code reviewers, we are expecting the following structure:
```   
 ...
    /src
        |---/modules
               | functionality.js
        | index.js
```
Please add some part of code to functionality.js file and import it in the index.js file as the second reviewer mentioned.

Keep rocking 🚀!

**Comment by Bestbynature** [comment]
@medaminedev66, Kindly point me to the part where you could find the requirement you pasted and sent to me. I believe you got them from a different project, not the current one. The current project is [to-do list add and remove](https://github.com/microverseinc/curriculum-javascript/blob/main/todo-list/m3_add_remove.md)

A screenshot of the requirements has been pasted below.

In addition to this, all my colleagues who submitted today didn't follow this procedure either.
Proof of the right requirement

**Comment by medaminedev66** [comment]
Hi @Bestbynature,

Please accept my apologies for any confusion. The name of your web app is `TodoList-webpack-ES6` and we have another project related to ES6. Upon reviewing your pull request again and based on the project requirements, I confirm that your project is complete and the user can add, remove, and edit tasks as expected.

However, there is still a requirement from the project that needs to be implemented. As per the project requirements, it is necessary to create a new file to implement the CRUD (Create, Update, Delete) functionalities. This is an essential requirement to ensure that the code follows ES6 and is easily maintainable in the future. Creating a separate file for the CRUD operations will help keep the code organized and modular, making it easier to debug and maintain.

As you have already learned how to do that in the "Awesome books - ES6" project, we expect you to follow the (ES6) and implement the same for this project. 

In conclusion, I request you to create a new file for the (add, remove and edit) operations and import it into the index.js file. Doing so will help ensure that the code is well-structured and maintainable, and follows (ES6).

Project requirement:
image


Keep rocking 🚀!

File: src/index.js [hunk 3]
```diff
@@ -1,36 +1,115 @@
...
+      });
+      if (test) {
+        abnormal.style.display = 'block';
+        abnormal.innerHTML = 'Sorry! You can\'t add the same task twice';
+      } else {
+        abnormal.style.display = 'none';
+        task.description = newValue;
+        store();
+      }
+    });
```
Comment (ABDULALI3468) [inline src/index.js]: As requested by the previous Code Reviewer its a project requirement to have this EDIT function in another file(Module) and import it here to use in index.js.

## Confused about how to do it???

The thing is you can use parameters in the function because the variables you are using here are will not be available in that file.

File: src/index.js [hunk 4]
```diff
@@ -1,36 +1,115 @@
...
+        tasks = newTasks.map((mapped, i) => ({
+          description: `${mapped.description}`,
+          index: `${i + 1}`,
+          completed: false,
+        }));
+        domTasks.innerHTML = '';
+        displayTasks();
+      }
+    });
   });
```
Comment (ABDULALI3468) [inline src/index.js]: As requested by the previous Code Reviewer it's a project requirement to have this **DELETE** function in another file(Module) and import it here to use in index.js.

## Confused about how to do it???

The thing is you can use parameters in the function because the variables you are using here are will not be available in that file.

**Review by ABDULALI3468 — CHANGES_REQUESTED** [review]
# Changes Required 👍 

Hi @Bestbynature , 



Great work on making the changes requested by a previous reviewer 👏 
You've done really well implementing some of the requested changes, but there are still some which aren't been addressed yet.


## Highlights 
✅All linters are checked.
✅Project is professionally documented.
✅The right GitHub flow method has been used.
✅Professional pull request title and description.


 ## Required Changes ♻️

> **_Check the comments below to have a look at the required changes._**

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

> File named functionaliy.js is having almost the same content as in index.js so please consider using the DRY(Don't Repeat Yourself) methodology here to make your code neat and tidy.

## Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear, and please remember to tag me in your question so I can receive the notification.

> * This is My **GitHub**: GitHub followers
> * This is My **Linkedin**: [Linkedin Badge](https://www.linkedin.com/in/abdul-ali-5400bb216/)&nbsp;
> * This is My **Twitter**: [Twitter Badge](https://twitter.com/mrabdul_ali)&nbsp;

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by mahbubul14 — APPROVED** [review]
# Project Approved :trophy: :tada: 

Hi @Bestbynature, 

Your project is complete!  You have done it exceptionally well. it's time to merge it :100: 
Congratulations! 🎉 

### To Highlight :+1: 

:heavy_check_mark: No linter error.
:heavy_check_mark: Descriptive readme file.
:heavy_check_mark: Well-implemented previous suggestions.

### Optional suggestions :arrow_down_small: 

Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better.

* N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification. You can also connect with me on [slack](https://microverse-students.slack.com/team/U01J39GL614)

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
```
